### PR TITLE
theme: add type for component overrides

### LINKED ISSUE #18422: 🐛 Bug Report: createUnifiedTheme does not allow overrides for Backstage components
### 📜 Description

Neither the v4 nor the v5 Theme currently supports overrides for non Material-UI components when created via the new createUnifiedTheme (https://github.com/backstage/backstage/blob/3f304f6211620ed9386d905c53822d46e574a48f/packages/theme/src/unified/UnifiedTheme.tsx#L73C27-L73C27) method.

v4 fails here since transformV5ComponentThemesToV4 (https://github.com/backstage/backstage/blob/3f304f6211620ed9386d905c53822d46e574a48f/packages/theme/src/unified/overrides.ts#L120C1-L123) filters out all entries that are not part of "ThemeOptions['components']" which only contains the Material-UI components.

For this to work the Backstage components would need to be added via module augmentation as seen here https://stackoverflow.com/questions/76095729/how-can-i-add-my-custom-component-to-mui-theme-options-so-that-i-can-override-it

### 👍 Expected behavior

Adding Backstage component overrides via the "components" option and using styleOverrides should apply those overrides to the Backstage component.

### 👎 Actual Behavior with Screenshots

The overrides are not applied and TypeScript is unhappy because it doesn't know what to do with the Backstage component names in the "components" section of the options.

### 👟 Reproduction steps

1. Create a custom theme via createUnifiedTheme.
2. Add some Backstage component to the "components" option.
3. Check the result.

### 📃 Provide the context for the Bug.

_No response_

### 🖥️ Your Environment

_No response_

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

None

### PR DESCRIPTION
**Pull request by Rugvip**
## Hey, I just made a Pull Request!

Fixes #18422. This introduces a way to globally register component override keys through module augmentation. Since we've only done style overrides I figured we'd stick to that for now. If we at some point want to allow for prop and variant overrides I figured we could add a separate type that allows that to be augmented. Was striving to keep it simple for now.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Review by freben — APPROVED** [review]
That's so fancy.

**Review by tudi2d — APPROVED** [review]
first of all: this is amazing - good job! do we maybe need tests & documentation for this?

**Comment by awanlin** [comment]
+1 to the suggestion of adding some docs, even if it's just something in this PR. I can run with that and up date the Aperture theme in the Demo site once this goes out in a `next` release and then we can point to that as a solid example 🚀

File: packages/theme/src/v5/types.ts [hunk 1]
```diff
@@ -19,6 +19,8 @@ import {
   BackstagePaletteAdditions,
   BackstageThemeAdditions,
 } from '../base/types';
+// eslint-disable-next-line no-restricted-imports
```
[the remark this answers is not in the document]
Comment (Rugvip) [reply packages/theme/src/v5/types.ts:22]: I think it's alright tbh still, but worth keeping an eye on. This is a fairly rare internal import and generally the things you need are re-exported from the parent module.

File: packages/theme/src/v5/types.ts [hunk 2]
```diff
@@ -35,3 +37,30 @@ declare module '@mui/material/styles' {
...
+ * ```
+ */
+export interface OverrideComponentNameToClassKeys {}
+
+type BackstageComponentOverrides = {
+  [TName in keyof OverrideComponentNameToClassKeys]?: {
+    styleOverrides?: Partial<
+      OverridesStyleRules<OverrideComponentNameToClassKeys[TName], TName, Theme>
+    >;
+  };
```
[the remark this answers is not in the document]
Comment (Rugvip) [reply packages/theme/src/v5/types.ts:61]: Added! 😁

**Comment by Rugvip** [comment]
Dunno if docs are needed here tbh, the theme creation part is kind of supposed to just work, and on the declaration end we don't have docs and tbh dunno if we should at this point

**Pull request merged**